src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs

```diff
@@ -343,16 +344,65 @@ internal static string DecompressToken(byte[] tokenBytes, string algorithm, int
-                throw LogHelper.LogExceptionMessage(new SecurityTokenDecryptionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10603, keysAttempted, (object)exceptionStrings ?? "", LogHelper.MarkAsSecurityArtifact(decryptionParameters.EncodedToken, SafeLogJwtToken))));
+            ExceptionDetail exceptionDetail = ValidateDecryption(
+                decryptionParameters,
+                decryptionSucceeded,
+                algorithmNotSupportedByCryptoProvider,
+                exceptionStrings,
+                keysAttempted,
+                null);
+
+            if (exceptionDetail is not null)
+                throw LogHelper.LogExceptionMessage(exceptionDetail.GetException());
+        }
 
-            if (!decryptionSucceeded && algorithmNotSupportedByCryptoProvider)
-                throw LogHelper.LogExceptionMessage(new SecurityTokenDecryptionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10619, LogHelper.MarkAsNonPII(decryptionParameters.Alg), LogHelper.MarkAsNonPII(decryptionParameters.Enc))));
+        private static ExceptionDetail ValidateDecryption(
+            JwtTokenDecryptionParameters decryptionParameters,
+            bool decryptionSucceeded,
+            bool algorithmNotSupportedByCryptoProvider,
+            StringBuilder exceptionStrings,
+            StringBuilder keysAttempted,
+#pragma warning disable CA1801 // Review unused parameters
+            CallContext callContext)
+#pragma warning restore CA1801 // Review unused parameters
+        {
+            ExceptionDetail exceptionDetail = null;
 
-            if (!decryptionSucceeded)
-                throw LogHelper.LogExceptionMessage(new SecurityTokenDecryptionFailedException(LogHelper.FormatInvariant(TokenLogMessages.IDX10609, LogHelper.MarkAsSecurityArtifact(decryptionParameters.EncodedToken, SafeLogJwtToken))));
+            if (!decryptionSucceeded && keysAttempted is not null)
+                return new ExceptionDetail(
+                    new MessageDetail(
+                        TokenLogMessages.IDX10603,
+                        keysAttempted.ToString(),
+                        exceptionStrings?.ToString() ?? string.Empty,
+                        LogHelper.MarkAsSecurityArtifact(decryptionParameters.EncodedToken, SafeLogJwtToken)),
+                    typeof(SecurityTokenDecryptionFailedException),
+                    new StackFrame(true),
+                    null);
+            else if (!decryptionSucceeded && algorithmNotSupportedByCryptoProvider)
+                return new ExceptionDetail(
+                    new MessageDetail(
+                        TokenLogMessages.IDX10619,
+                        LogHelper.MarkAsNonPII(decryptionParameters.Alg),
+                        LogHelper.MarkAsNonPII(decryptionParameters.Enc)),
+                    typeof(SecurityTokenDecryptionFailedException),
+                    new StackFrame(true),
+                    null);
+            else if (!decryptionSucceeded)
```

Comment:
this could become a more general decryption failure case w/just an `else`

Reply:
Refactored.